### PR TITLE
MassaLabs: Eliminate exponentiations in the IR

### PR DESCRIPTION
**Pull request by Leo-Besancon**
This PR completes https://github.com/0xPolygonMiden/air-script/issues/333, and removes the Operation::Exp node in the IR. Instead, parsed exponentiations are reduced to a series of multiplications using square and multiply.

### CODE REVIEW

**Comment by bitwalker** [comment]
@Leo-Besancon I've addressed the clippy warnings in `next` with the merge of #351, however we have some conflicts, so you'll have to rebase. I don't think there's actually any significant conflicts, just overlapping disjoint changes in the same files, but I haven't checked. Let me know if you have any questions about how to resolve any conflicts you aren't sure about.

**Review by bobbinth — COMMENTED** [review]
Looks great! Thank you! My only one suggestion would be to modify one of the tests somewhere so that we see how expanding something other than squaring works (e.g., have `x^7` somewhere).

**Comment by Leo-Besancon** [comment]
Thank you @bitwalker and @bobbinth, I resolved the small conflicts and updated the complex function test to test the 7 case.

File: air-script/tests/functions/functions_complex.rs [hunk 1]
```diff
@@ -86,6 +86,6 @@ impl Air for FunctionsAir {
...
-        result[0] = aux_next[0] - aux_current[0] * ((E::from(main_current[3]).exp(E::PositiveInteger::from(4_u64)) * E::from(main_current[1]) * E::from(main_current[2]) + E::from(main_current[3]).exp(E::PositiveInteger::from(2_u64)) * (E::ONE - E::from(main_current[1])) * E::from(main_current[2]) + E::from(main_current[3]) * E::from(main_current[1]) * (E::ONE - E::from(main_current[2])) + (E::ONE - E::from(main_current[1])) * (E::ONE - E::from(main_current[2]))) * E::from(main_current[0]) - E::from(main_current[0]) + E::ONE);
+        result[0] = aux_next[0] - aux_current[0] * ((E::from(main_current[3]) * E::from(main_current[3]) * E::from(main_current[3]) * E::from(main_current[3]) * E::from(main_current[3]) * E::from(main_current[3]) * E::from(main_current[3]) * E::from(main_current[1]) * E::from(main_current[2]) + E::from(main_current[3]) * E::from(main_current[3]) * (E::ONE - E::from(main_current[1])) * E::from(main_current[2]) + E::from(main_current[3]) * E::from(main_current[1]) * (E::ONE - E::from(main_current[2])) + (E::ONE - E::from(main_current[1])) * (E::ONE - E::from(main_current[2]))) * E::from(main_current[0]) - E::from(main_current[0]) + E::ONE);
```
Comment (bobbinth) [inline air-script/tests/functions/functions_complex.rs:89]: Not for this PR, but I think the ideal output here would be something like:
```rust
let tmp1 = main_current[3] * main_current[3]; // main_current[3]^2
let tmp2 = tmp1 * tmp1;                       // main_current[3]^4
let tmp3 = F::ONE - main_current[1];
let tmp4 = F::ONE - main_current[2];

result[0] = aux_next[0] - aux_current[0] * E::from(
    (
        main_current[3] * tmp2 * tmp1 * main_current[1] * main_current[2]
        + tmp2 * tmp3 * main_current[2]
        + main_current[3] * main_current[1] * tmp4 
        + tmp3 * tmp4
    )
    * main_current[0] - main_current[0] + F::ONE
);
```
The main points being:
- We introduce temporary variables to hold results of common sub-expressions.
- We delay conversion to the extension filed to the end.

**Review by bobbinth — APPROVED** [review]
Looks good! Thank you! I left one comment inline - but this is more for the time when we work on the codegen side. Could you create an issue based on it?

File: air-script/tests/functions/functions_complex.rs (hunk 1 above)
Comment (Leo-Besancon) [reply air-script/tests/functions/functions_complex.rs:89]: Issue created here: https://github.com/0xPolygonMiden/air-script/issues/354

**Pull request merged**